finetune/input_pipeline.py

```diff
@@ -0,0 +1,209 @@
+import itertools
+import logging
+import sys
+
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.data import Dataset
+
+from finetune.config import PAD_TOKEN
+from finetune.encoding import TextEncoder, ArrayEncodedOutput, EncodedOutput
+from finetune.imbalance import compute_class_weights
+
+
+class BasePipeline(metaclass=ABCMeta):
+    def __init__(self, config):
+        self.config = config
+        self.label_encoder = None
+        self.encoder = TextEncoder()
+        self.target_dim = None
+        self.pad_idx_ = None
+        self.rebuild = False
+
+    @abstractmethod
+    def _target_encoder(self):
+        # Overridden by subclass to produce the right target encoding for a given target model.
+        raise NotImplementedError
+
+    def feed_shape_type_def(self):
+        TS = tf.TensorShape
+        return ({"tokens": tf.int32, "mask": tf.float32}, tf.float32), (
+            {"tokens": TS([self.config.max_length, 2]), "mask": TS([self.config.max_length])}, TS([self.target_dim]))
+
+    def _array_format(self, encoded_output, pad_token=PAD_TOKEN):
+        """
+        Returns numpy array of token idxs and corresponding mask
+        Returned `x` array contains two channels:
+            0: byte-pair encoding embedding
+            1: positional embedding
+        """
+        seq_length = len(encoded_output.token_ids)
+        x = np.zeros((self.config.max_length, 2), dtype=np.int32)
+        mask = np.zeros((self.config.max_length), dtype=np.float32)
+
+        if encoded_output.labels is not None:
+            labels_arr = np.empty((self.config.max_length), dtype='object')
+            labels_arr.fill(pad_token)
+        else:
+            labels_arr = None
+
+        # BPE embedding
+        x[:seq_length, 0] = encoded_output.token_ids
+        # masking: value of 1 means "consider this in cross-entropy LM loss"
+        mask[1:seq_length] = 1
+        if encoded_output.labels:
+            labels_arr[:seq_length] = encoded_output.labels
+        # positional_embeddings
+        x[:, 1] = np.arange(self.encoder.vocab_size, self.encoder.vocab_size + self.config.max_length)
+
+        return ArrayEncodedOutput(
+            token_ids=x,
+            tokens=encoded_output.tokens,
+            labels=labels_arr,
+            char_locs=encoded_output.char_locs,
+            mask=mask,
+        )
+
+    def _adapt_feats_for_distributed(self, feats_labels):
+        if type(feats_labels) == tuple:
+            feats, labels = feats_labels
+            feats["labels"] = labels
+            return feats
+
+    def text_to_tokens_mask(self, X, Y=None):
+        out_gen = self._text_to_ids(X)
+        for out in out_gen:
+            feats = {"tokens": out.token_ids, "mask": out.mask}
+            if Y is None:
+                yield feats
```

Comment:
I think it's good in general to shy away from functions that have optional return values.  Leads to gross logic in IMO.  Functions should always have consistent return types.